sdk/storage/src/data_lake/clients/data_lake_client.rs

```diff
@@ -10,45 +13,63 @@ use url::Url;
 const DEFAULT_DNS_SUFFIX: &str = "dfs.core.windows.net";
 
 pub trait AsDataLakeClient<A: Into<String>> {
-    fn as_data_lake_client(&self, account: A) -> Result<Arc<DataLakeClient>, url::ParseError>;
+    fn as_data_lake_client(
+        &self,
+        account: A,
+        bearer_token: String,
+    ) -> Result<Arc<DataLakeClient>, url::ParseError>;
 }
 
 pub trait AsCustomDataLakeClient<DS: Into<String>, A: Into<String>> {
     fn as_data_lake_client_with_custom_dns_suffix(
         &self,
         account: A,
+        bearer_token: String,
         dns_suffix: DS,
     ) -> Result<Arc<DataLakeClient>, url::ParseError>;
 }
 
 impl<A: Into<String>> AsDataLakeClient<A> for Arc<StorageClient> {
-    fn as_data_lake_client(&self, account: A) -> Result<Arc<DataLakeClient>, url::ParseError> {
-        DataLakeClient::new(self.clone(), account.into(), None)
+    fn as_data_lake_client(
+        &self,
+        account: A,
+        bearer_token: String,
```

Comment:
@heaths

Reply:
All our SDKs take an endpoint and a credential, which is almost always a `TokenCredential` define in azure_core, with implementations like `DefaultAzureCredential` in azure_identity. This is laid out in our [guidelines](https://azure.github.io/azure-sdk/general_introduction.html).

I recommend taking a look at what we've done for our other Azure SDKs - any packages starting with "azure*" (as opposed to "microsoft.azure*" or something). I'm not sure about data lake specifically, but for storage in general we already have good examples you should match (type names, method names besides casing, param names, etc. - something we all do across official languages):

For example,

* https://github.com/Azure/azure-sdk-for-net/tree/main/sdk/storage/Azure.Storage.Files.DataLake/src
* https://github.com/Azure/azure-sdk-for-java/tree/main/sdk/storage/azure-storage-file-datalake/src

Specific to authentication, the `TokenCredential` is responsible for authenticating the user and getting a token. How that is attached to your request is up to you (and we have some helps in azure_core that may help, though not as many built out as with our other languages). In general, you should not roll your own authentication. Some libraries will, in addition, take a connection string. Follow whatever official languages are doing for guidance in those cases, though.

Reply:
Thanks @heaths. @rylev, what is the approach for something like this? Should all new code use this design, or should we track an issue where we switch over everything at once?

Reply:
Ok, I've decided to do the TokenCredential refactoring as a separate PR either way.